packages/tupaia-web/src/features/Matrix.tsx

```diff
@@ -0,0 +1,114 @@
+  Alert,
+} from '@tupaia/ui-components';
+import { MatrixDataColumn, MatrixDataRow, MatrixViewContent } from '../types';
+import { ConditionalPresentationOptions } from '@tupaia/types';
+import styled from 'styled-components';
+
+const NoDataMessage = styled(Alert).attrs({
+  severity: 'info',
+})`
+  width: 100%;
+  margin: 1rem auto;
+  max-width: 24rem;
+`;
+
+// This is a recursive function that parses the rows of the matrix into a format that the Matrix component can use.
+const parseRows = (
+  rows: MatrixDataRow[],
+  categoryId?: MatrixDataRow['categoryId'],
+): MatrixRowType[] => {
+  let topLevelRows = [];
+  // if a categoryId is not passed in, then we need to find the top level rows
+  if (!categoryId) {
+    // get the highest level rows, which are the ones that have a category but no categoryId
+    const highestLevel = rows.filter(row => row.category && !row.categoryId);
+    // if there are no highest level rows, then the top level rows are just all of the rows
+    topLevelRows = highestLevel.length ? highestLevel : rows;
+  } else {
+    // otherwise, the top level rows are the ones that have the categoryId that was passed in
+    topLevelRows = rows.filter(row => row.categoryId === categoryId);
+  }
+
+  // loop through the topLevelRows, and parse them into the format that the Matrix component can use
+  return topLevelRows.map(row => {
+    const { dataElement = '', category, ...rest } = row;
+    // if the row has a category, then it has children, so we need to parse them using this same function
+    if (category) {
+      return {
+        title: category,
+        ...rest,
+        children: parseRows(rows, category),
+      };
+    }
+    // otherwise, handle as a regular row
+    return {
+      title: dataElement,
+      ...rest,
+    };
+  });
+};
+
+// This is a recursive function that parses the columns of the matrix into a format that the Matrix component can use.
+const parseColumns = (columns: MatrixDataColumn[]): MatrixColumnType[] => {
+  return columns.map(column => {
+    const { category, key, title, columns: children } = column;
+    // if a column has a category, then it has children, so we need to parse them using this same function
+    if (category)
+      return {
+        title: category,
+        key: category,
+        children: parseColumns(children!),
+      };
+    // otherwise, handle as a regular column
+    return {
+      title,
+      key,
+    };
+  });
+};
+
+/**
+ * This is the component that is used to display a matrix. It handles the parsing of the data into the format that the Matrix component can use, as well as placeholder images. It shows a message when there are no rows available to display.
+ */
+
+interface MatrixProps {
+  viewContent: MatrixViewContent;
+  isEnlarged?: boolean;
+}
+export const Matrix = ({ viewContent, isEnlarged = false }: MatrixProps) => {
+  const { columns, rows, ...config } = viewContent;
+
+  const getPlaceholderImage = () => {
```

Comment:
This is very minor but I reckon this would be easier to read if it was placed outside of the Matrix component just above it and just took config or presentationOptions and categoryPresentationOptions as props